app/app/src/main/java/com/example/myapplication/MainActivity.java

```diff
@@ -410,6 +423,52 @@ void goBackward() throws IOException {
         int msg = 5;
         Car.mmOutputStream.write(msg);
     }
+
+    // joyStick control methods to replace four buttons
+    public void onMove(JoyStick joyStick, double angle, double power, int direction) {
```

Comment:
Before allowing move you have to use the following toast if the car is not connected to the app.

`private void pleaseConnectCar() {
        String toastString = "Please connect to smart car";
        Toast.makeText(getApplicationContext(), toastString, Toast.LENGTH_SHORT).show();
    }`

Reply:
Is there a way of only having the toast display once? Because having it directly inside of the switch case in onMove makes it display every time there is a move, and it takes a long time for it to disappear that way.

Reply:
This fix can be a separate "bug" labeled issue once the current branch is merged to master :) it might not be obvious how to fix now.